### DISCUSSION TITLE
Add resource directly from host

### DISCUSSION THREAD

**Discussion by disl-pdavid** [ckan/ckan#6985] [Q&A] [answered]
Hello again!

I'm curious - is there a way to manually add a resource to a dataset directly from the host machine?

I have some datasets that are a little large (~3GB) and I didn't want to waste time on uploading through the web interface when the data already resides on the same host machine running CKAN.  I checked the commands locally but nothing stood out as addressing adding a resource.

**Comment by wardi** [comment, marked as answer]
`ckanapi -c $CKAN_INI action resource_patch id=$RESOURCE_ID upload@$FILE_LOCATION` should work

**Comment by disl-pdavid** [comment]
@wardi thanks!  This presumes to be patching an existing resource, right?  How does this change to upload a brand new resource and adding it to an existing package?  I've tried `resource_create` and it appears to work but the resource is not given an UUID and there is no download of the resource available on the dataset page.

**Comment by wardi** [comment]
it should work the same with `resource_create`, you just need to pass the `package_id` instead of a `resource_id`

**Comment by disl-pdavid** [comment]
Using
```
ckanapi -c /etc/ckan/production.ini action resource_create package_id='76b6b964-22a8-4327-83fe-a16f51ebf76c' name='test.txt' upload@test.txt
```

It _appears_ to work and returns:
```
{
  "cache_last_updated": null,
  "cache_url": null,
  "created": "2022-07-26T16:00:39.460141",
  "datastore_active": false,
  "datastore_contains_all_records_of_source_file": false,
  "description": null,
  "format": "",
  "hash": "",
  "id": "14a9866c-ce54-459c-8c3f-8f0e5996b207",
  "last_modified": null,
  "metadata_modified": "2022-07-26T16:00:39.425058",
  "mimetype": null,
  "mimetype_inner": null,
  "name": "test.txt",
  "package_id": "76b6b964-22a8-4327-83fe-a16f51ebf76c",
  "position": 8,
  "resource_type": null,
  "size": null,
  "state": "active",
  "url": "",
  "url_type": null
}
```

But notice it doesn't create any sort of URL or resource_id - should these have been generated on creation?

**Reply by wardi** [reply]
`id` is the resource_id, but it's strange that there's no url shown. Try creating the resource then uploading it afterwards with package_patch. Or try the same call over http (using `-r` and an API key/token), or try without passing a `name` value and see if it's any different.

**Reply by disl-pdavid** [reply]
So, I actually call an ingest python script and use the http API endpoints to create a package and then upload resources tied to the package.  This all works great and the resources are visible and can be downloaded no problem (all works as expected).

### Using http `resource_create` endpoint:
```
{
      "cache_last_updated": null,
      "cache_url": null,
      "ckan_url": "https://mydomain",
      "created": "2022-07-26T15:57:49.049540",
      "datastore_active": true,
      "datastore_contains_all_records_of_source_file": true,
      "description": null,
      "format": "CSV",
      "hash": "29617876861c4362ba97e5921a8fbec2",
      "id": "1addf7c4-75f3-47cc-9121-0706ceec861a",
      "ignore_hash": false,
      "last_modified": "2022-07-26T15:57:48.972061",
      "metadata_modified": "2022-07-26T15:57:56.142087",
      "mimetype": "text/csv",
      "mimetype_inner": null,
      "name": "mangrove_bg_production.csv",
      "original_url": "https://mydomain/dataset/76b6b964-22a8-4327-83fe-a16f51ebf76c/resource/1addf7c4-75f3-47cc-9121-0706ceec861a/download/mangrove_bg_production.csv",
      "package_id": "76b6b964-22a8-4327-83fe-a16f51ebf76c",
      "position": 7,
      "resource_id": "1addf7c4-75f3-47cc-9121-0706ceec861a",
      "resource_type": null,
      "set_url_type": false,
      "size": 1575,
      "state": "active",
      "task_created": "2022-07-26 15:57:49.312794",
      "url": "https://mydomain/dataset/76b6b964-22a8-4327-83fe-a16f51ebf76c/resource/1addf7c4-75f3-47cc-9121-0706ceec861a/download/mangrove_bg_production.csv",
      "url_type": "upload"
    },
    {
      "cache_last_updated": null,
      "cache_url": null,
      "created": "2022-07-26T16:00:39.460141",
      "datastore_active": false,
      "datastore_contains_all_records_of_source_file": false,
      "description": null,
      "format": "",
      "hash": "",
      "id": "14a9866c-ce54-459c-8c3f-8f0e5996b207",
      "last_modified": null,
      "metadata_modified": "2022-07-26T19:29:35.886272",
      "mimetype": null,
      "mimetype_inner": null,
      "name": "test.txt",
      "package_id": "76b6b964-22a8-4327-83fe-a16f51ebf76c",
      "position": 8,
      "resource_type": null,
      "size": null,
      "state": "active",
      "url": "",
      "url_type": null
    }

```
That first entry (mangrove_bg_production.csv) is one created using the http API `resource_create` endpoint.  Works as expected.  The second one (test.txt) is using ckanapi `action resource_create` with the same params (package id, name, file upload).

### Passing w/o a name
Trying it without specifying a name doesn't change behavior. :(

Basically, it all works as expected when POSTing through the http API using python or curl.  Using ckanapi doesn't create the resource correctly (it doesn't exist in `/var/lib/ckan/resources`)

**Reply by disl-pdavid** [reply]
Not sure if related, but calling the `dataset_purge` action with:

```
$ curl -H"Authorization: ${APITOKEN}" -X POST https://domain/api/action/dataset_purge -k -d "{\"id\":\"3129a032-bc1a-4049-92b3-bca153385df5\" }"

{"help": "https://domain/api/3/action/help_show?name=dataset_purge", "success": true, "result": null}
```

also trips an error in my logs:

```
ckan     | 2022-07-26 20:17:21,107 ERROR [ckan.model.modification] 
ckan     | Traceback (most recent call last):
ckan     |   File "/usr/lib/ckan/venv/src/ckan/ckan/model/modification.py", line 67, in notify
ckan     |     observer.notify(entity, operation)
ckan     |   File "/usr/lib/ckan/venv/src/ckan/ckan/lib/search/__init__.py", line 168, in notify
ckan     |     logic.get_action('package_show')(cast(
ckan     |   File "/usr/lib/ckan/venv/src/ckan/ckan/logic/__init__.py", line 531, in wrapped
ckan     |     result = _action(context, data_dict, **kw)
ckan     |   File "/usr/lib/ckan/venv/src/ckan/ckan/logic/action/get.py", line 999, in package_show
ckan     |     raise NotFound
ckan     | ckan.logic.NotFound

```
